GPU/Common/TextureCacheCommon.cpp

```diff
@@ -2066,6 +2152,128 @@ void TextureCacheCommon::ApplyTextureFramebuffer(VirtualFramebuffer *framebuffer
 	gstate_c.Dirty(DIRTY_BLEND_STATE | DIRTY_DEPTHSTENCIL_STATE | DIRTY_RASTER_STATE | DIRTY_VIEWPORTSCISSOR_STATE);
 }
 
+// Applies depal to a normal (non-framebuffer) texture, pre-decoded to CLUT8 format.
+void TextureCacheCommon::ApplyTextureDepal(TexCacheEntry *entry) {
+	Draw2DPipeline *textureShader = nullptr;
+	uint32_t clutMode = gstate.clutformat & 0xFFFFFF;
+
+	switch (entry->format) {
+	case GE_TFMT_CLUT4:
+	case GE_TFMT_CLUT8:
+		break; // These are OK
+	default:
+		_dbg_assert_(false);
+		return;
+	}
+
+	// Create GPU resources.
+	if (!dynamicClutFbo_) {
+		Draw::FramebufferDesc desc{};
+		desc.width = 512;
+		desc.height = 1;
+		desc.depth = 1;
+		desc.z_stencil = false;
+		desc.numColorAttachments = 1;
+		desc.tag = "dynamic_clut";
+		dynamicClutFbo_ = draw_->CreateFramebuffer(desc);
+		dynamicClutReinterpreted_ = draw_->CreateFramebuffer(desc);
+	}
+
+	const GEPaletteFormat clutFormat = gstate.getClutPaletteFormat();
+	u32 depthUpperBits = 0;
+
+	// The CLUT texture is dynamic, it's the framebuffer pointed to by clutRenderAddress.
+	// Instead of texturing directly from that, we copy to a temporary CLUT texture.
+	GEBufferFormat expectedCLUTBufferFormat = (GEBufferFormat)clutFormat;  // All entries from clutFormat correspond directly to buffer formats.
+
+	VirtualFramebuffer *src = framebufferManager_->GetVFBAt(clutRenderAddress_);
+	if (!src) {
+		// What do we do?
+		return;
+	}
+
+	Draw::Framebuffer *clutFbo = dynamicClutFbo_;
+
+	// First we use a blit (with nearest interpolation, so we don't mash pixels together)
+	// to shrink to the correct size, if we are running with scaling.
+	// We can always blit 512 pixels even if we only need less, the cost will be negligible.
+	framebufferManager_->BlitUsingRaster(
```

Comment:
Hm, ideally we'd do this on loadclut.  I could totally see a game rendering a clut, loading it, and then rendering straight away on top of the rendered clut.  In fact, Brave Story might even do this?  I don't remember... well, probably not.

-[Unknown]

Reply:
Hm, there's one issue here, I think we then have to always copy to the temp here, and do the reinterpret (if needed) at the time of ApplyTextureDepal, since the specified clut format can change in between.

That does sound reasonable though, I will take care of that later. It's just an extra tiny copy in the case of matching formats.